### DISCUSSION TITLE
We're creating a maintainer team! Join us :D

### DISCUSSION THREAD

**Discussion by willingc** [jupyter/nbgrader#1637] [Announcements]
Hi all,

We're hoping some of our long-time users or enthusiasts for nbgrader would like to be part of a team of individuals maintaining nbgrader.

We've structured JupyterHub and Binder in a similar way. See https://jupyterhub-team-compass.readthedocs.io/en/latest/team.html#jupyterhub-team 

Please leave a comment on this issue. If you have questions, please leave them here too.

**Comment by perllaghu** [comment]
Can I propose @BertR ?
(Lead developer for the Noteable service run out of The University of Edinburgh, and done a lot of work with the external exchange plugin, as well as other parts of nbgrader)

**Comment by willingc** [comment]
@perllaghu We're going to email those that participated in the sprint last year. I agree @BertR would be a great addition to the maintainer team. Looking forward to hearing back from @BertR. :sunny:

**Comment by BertR** [comment]
Hi everyone, I would like to join the mainainer team. Thanks for the suggestion @perllaghu and for setting this in motion @willingc

**Comment by jhamrick** [comment]
Thanks so much for opening this @willingc !

First of all, I really, sincerely apologize for dropping the ball on responding to emails, issues, and PRs---life has caught up with me and I have been really struggling to find extra time to look at nbgrader stuff recently 😞

I hope this will mean that the PRs that you all have been making (and that I have been failing to review) will be able to move forward. We will grant everyone in the maintainer's team the ability to do standard maintainer stuff on the repository: merge PRs, triage (close and tag) issues, etc. so this should make it easier for all of you to move forward on the improvements and changes that you need for your setups.

Please do let us know if you have any questions!

**Comment by jhamrick** [comment]
@BertR I have gone ahead and added you---thanks a lot!

@perllaghu would you like to be added to the maintainers team too?

**Comment by willingc** [comment]
No worries @jhamrick. I'm in the same time boat as you. Miss seeing you in person too. Stay safe.

**Comment by perllaghu** [comment]
> @perllaghu would you like to be added to the maintainers team too?

I think just one from EDINA is enough :)

**Comment by nthiery** [comment]
Thanks for the invitation. Happy to help if I can, though I have limited time and expertise of nbgrader.

**Comment by dsblank** [comment]
Thanks also for the invitation. I keep an eye on the repo, but have retired from teaching. Feel free to reach out to me if there is ever anything I can do to help!

**Comment by damianavila** [comment]
@jhamrick count me in as part of the team. I will try to help as long as I can. Cheers.

**Comment by choldgraf** [comment]
hey @jhamrick and @willingc - I am happy to be a part of the maintainer team, and will do what I can to help out.

**Comment by rkdarst** [comment]
Thanks for the invitation!  I think that right now, I should avoid other responsibilities since my time is too limited and nbgrader "mostly works for me" these days so that it's not one of my main focuses.  Still, I could do some useful stuff sometimes - summers are the Jupyter development sprints, so expect more around then.

I've been thinking of what it takes to be on a maintainer team, and it's more reviewing and accepting other people's changes: being able to take risks, and fix stuff when it goes wrong.  By this standard, too, I think that other people who have mental energy to put to this are better for the team.  My wish to you all: be bold and take risks, we can debug and fix together.

**Comment by willingc** [comment]
If there are no objections, I will add: @choldgraf @damianavila to the maintainer team. Since @dsblank @rkdarst @nthiery @perllaghu have all been long-time supporters of nbgrader, I will increase your privileges triage team so that you can label issues. Thanks everyone.

**Comment by rkdarst** [comment]
I was wondering, what are everyone's nbgrader plans?  I am about to talk to people in my department about what we should do in the future.  Depending on interest, I will advocate for hiring a developer to push forward on various things.  In the last years I have seen many ideas for development (jupyterlab, etc) but not much has happened.

Should we arrange some sort of quick check-in meeting?  Is it better to post on discourse?  New issue?

**Comment by perllaghu** [comment]
I don't know about good/bad/right places.... I know I wouldn't see it if it was on discourse

I also don't know who to pull in/allow in: I can see asking too open a question would result in 23,562 requests for changes, with 0.01hrs of offered help.... just like this answer :D 

There's no end of requests for features & things - however I can list the things that are of interest to me:
* Continue the `jupyterlab` UI work
* Review/consider the multiple-marker options [bearing in mind we now have _pluggable exchanges_ available]
* Get some implementation descriptions to help those who are looking to use nbgrader outside the _1 marker 1 course 1 computer_ environment... in particular those places that are spawning notebooks into some cloud/cluster environment - eg:
> "We are using AWS, where our service is running in a virtual host. AWS allows us to expand/contract the resources for that host [CPU count, RAM, file-store] as needed. Jupyterhub is using the DockerSpawner to spawn our pre-defined Classic Notebook. We have a custom authenticator for jupyterhub that uses an API for our central department servers"
* Tweak the `assignment_list` extension to only show the `Fetch Feedback` button when there's actually feedback that can be fetched
* ... and what's actually of interest to _your department_ and/or _your staff_ (given you [plural/abstract] are are paying for this)

**Comment by rkdarst** [comment]
In particular, I have seen various threads of people working on other aspects: JupyterLab, etc.  We need to see what the status of these projects are, to avoid duplication.  Perhaps it is in the issues, but we need more...

**Comment by ocordes** [comment]
Hi,

I would like to join the maintaining and also the developer team. I'm a developer and have some experience in programming of jupyterlab extensions. From the other side I'm a lecturer for python and similar topics and we tried to use nbgrader for lectures in a kubernetes installation. The experiences were good but we found several small issues which we want to solve. I have also applied successfully for a developer position in our university (Bonn, Germany) and we want to spend some effort for the development (however we need to look for a suitable person ;-))!

Anyway, one proof of concept for a jupyterlab-extension is maybe here: https://github.com/ocordes/jupyterlab_nbgrader which I started in May 2021 but I was stopped by the flood catastrophe in germany in July 2021. Now I can spend a little more time on this project.  

All the best for you and your families,
  Oliver

**Comment by BertR** [comment]
> Hi,
> 
> I would like to join the maintaining and also the developer team. I'm a developer and have some experience in programming of jupyterlab extensions. From the other side I'm a lecturer for python and similar topics and we tried to use nbgrader for lectures in a kubernetes installation. The experiences were good but we found several small issues which we want to solve. I have also applied successfully for a developer position in our university (Bonn, Germany) and we want to spend some effort for the development (however we need to look for a suitable person ;-))!
> 
> Anyway, one proof of concept for a jupyterlab-extension is maybe here: https://github.com/ocordes/jupyterlab_nbgrader which I started in May 2021 but I was stopped by the flood catastrophe in germany in July 2021. Now I can spend a little more time on this project.
> 
> All the best for you and your families, Oliver

Hi Oliver! That's great news. There are a couple of teams I know are working on making JupyterLab compatible
nbgrader happen. @perllaghu (University of Edinburgh), @jgwerner (IllumiDesk). Previously, some work was done by students at UC Davis, they wrote about their experiences here: https://ngshare.readthedocs.io/en/ecs193-final-report/.

It would be great if you could get together and setup a meeting.

**Comment by jgwerner** [comment]
> > Hi,
> 
> > 
> 
> > I would like to join the maintaining and also the developer team. I'm a developer and have some experience in programming of jupyterlab extensions. From the other side I'm a lecturer for python and similar topics and we tried to use nbgrader for lectures in a kubernetes installation. The experiences were good but we found several small issues which we want to solve. I have also applied successfully for a developer position in our university (Bonn, Germany) and we want to spend some effort for the development (however we need to look for a suitable person ;-))!
> 
> > 
> 
> > Anyway, one proof of concept for a jupyterlab-extension is maybe here: https://github.com/ocordes/jupyterlab_nbgrader which I started in May 2021 but I was stopped by the flood catastrophe in germany in July 2021. Now I can spend a little more time on this project.
> 
> > 
> 
> > All the best for you and your families, Oliver
> 
> 
> 
> Hi Oliver! That's great news. There are a couple of teams I know are working on making JupyterLab compatible
> 
> nbgrader happen. @perllaghu (University of Edinburgh), @jgwerner (IllumiDesk). Previously, some work was done by students at UC Davis, they wrote about their experiences here: https://ngshare.readthedocs.io/en/ecs193-final-report/.
> 
> 
> 
> It would be great if you could get together and setup a meeting.
> 
> 

@BertR @ocordes would love to participate in a meet as well, we have made quite a few updates lately to nbgrader but need to coordinate how to send these updates as PRs to this repo. We have ported the Formgrader app to a React based frontend and have refactored quite a bit of code / tests. We also are managing nbgrader autograder services as jobs with Argo events and Argo workflows but that's probably out of the scope of this repo.

As a side not we think that it's more efficient to merge nbexchange to use nbexchange's setup as the default (we have struggled with NFS permissions and such) - but that of course is just an opinion 😊.

**Comment by rkdarst** [comment]
If there is any kind of discussion about future of nbgrader maintenance, I would be interested in attending, too.  I haven't had time to do much myself, but we are trying to hire someone that can do some work on it.  And I have a big to-do list and all, and would like to help anyone else who could use it.

**Comment by damianavila** [comment]
Hey folks, you should be all aware there is some upcoming work in January by QuantStack's folks into porting nbgrader to jupyterlab: https://github.com/jupyter/enhancement-proposals/pull/79#issuecomment-988055233.
 
cc'ing @SylvainCorlay here so they are aware there is also interest here in that transition (among other things).

**Comment by SylvainCorlay** [comment]
Hey @damianavila thank you for the ping! Indeed we had a quote signed a few days ago for this work. We will communicate about this small grant soon.

**Comment by SylvainCorlay** [comment]
Here we go! QuantStack's work on nbgrader will be funded as part of a joint grant program between **Université de Paris** and the **National University of Singapore (NUS)**.

The official announcement listing the awarded proposals is available here: https://u-paris.fr/resultats-de-lappel-a-projets-universite-de-paris-nus-2021/.

We should be good to start on this project as soon as early 2022.

**Comment by brichet** [comment]
Porting notebook extensions to labextensions and make it compatible with Notebook 7 (and Jupyterlab) will break compatibility with classic Notebook.
According to the last [Notebook weekly meeting](https://github.com/jupyter/notebook-team-compass/issues/5#issuecomment-1109766781) the plan could be to :
- move the current master to a 0.6.x version, which will be compatible with classic notebook's extension system
- use master branch for labextensions development

**Comment by jtpio** [comment]
Thanks @brichet!

I'm happy to help setup the branches and prepare the repo to be able to iterate on new developments, and keep the `0.6.x` branch alive to maintain the latest stable release of `nbgrader`.

Maybe we can do a quick check at the notebook weekly meeting tomorrow again, in case anyone has any objection: https://github.com/jupyter/notebook-team-compass#weekly-team-meetings

Looking forward to it!

**Comment by BertR** [comment]
Happy to hear about the plans. I've been trying to get the test suite passing again, there's something going on within the selenium tests. I'll try to get that done this week, but let me know if there's other stuff I can help with.

**Comment by rkdarst** [comment]
This isn't the right issue but since we are talking about "who is doing what"... someone working with me has gotten things working with nbconvert v6 and most tests passing (except Windows, which I guess needs to be checked in the gh-actions output some):
 
https://github.com/AaltoSciComp/nbgrader/pull/7

I plan on cleaning up this PR or Friday (co-working to teach a new person) and then submitting a proper PR (that one is internal to our repo and has other unrelated stuff in it, it can't be used as-is)

**Comment by brichet** [comment]
@rkdarst this is good news because it is probably the first step to be taken.

**Comment by WolfgangFahl** [comment]
This issue could go to dicussions see #1627 - then it would not look like an open issue.
And yes - i'd love to join the maintainer team.

**Comment by rkdarst** [comment]
Hello all!

I was wondering what the current maintainer team status was.  We've got someone new starting with us at @AaltoSciComp, and are generating different PRs and improvements.  It would be good to know our status and all - I can help review pull requests or even merge them, but it would be good to have an outsider too.

What's the overall feeling of the development status?  Is it still moving forward?  Are we stuck in reviewing, since we no longer are sure what will or won't break things for others?  Or should we be more bold.  This is what I'm thinking about now...
